### ISSUE TITLE
[Feature] Asynchronous processing of keywords

### ISSUE THREAD

**Issue by olivierobert** [nhantran3395/scrapper#44]
## Issue

Upon uploading keywords, the scraping will be performed synchronously in the handler:

https://github.com/nhantran3395/scrapper/blob/448ebb7f173de08329614fbc14c41673c7099651/apps/api/src/routes/uploads/handlers/create-upload.handler.ts#L37-L45

The request can take a long time to complete (and potentially could exceed the HTTP request timeout). In addition, if any errors occur for any keyword, the processing of keywords will stop.

## Expected

- Keywords should be stored in the database with a flag to track the scraping status.
- A distinct background job to scrap the Google page is then scheduled.

The benefits are:

- Fast CSV upload request
- The scraping of each keyword is isolated, i.e., one could error out while others could be successful.

> [!Warning]
> Asynchronous processing of keywords was the major technical challenge in the application. It seems to have been missed on the project planning: https://github.com/nhantran3395/scrapper/wiki/%5BDeprecated%5D-The-main-challenges-and-approaches

**Comment by nhantran3395** [comment]
Hi @olivierobert : It is great that you caught this out. The upload processing is slow at the moment.

I have considered offloading the scrape processing (especially the synchronous operations of parsing each page content, when the number of keywords is high) with Node.js built-in worker threads, using the `workerpool` NPM library. Doing this could help with not blocking the server from handling other incoming requests. This was however not included in my last submission.

Following the suggestion you provided would be the best method, not only freeing the server so that it serves other requests but also reducing the wait time during uploading. I will work on this and raise a PR.

In order to keep it simple but still be able to update the client app once the processing is finished, is it okay if I add a timer to refresh the `uploads` status every X minute? (calling the `GET /uploads` endpoint every X min).

If a real-time status update is not required, I think loading the `uploads` list every X minute should be okay. We can start with 1 min.

**Comment by olivierobert** [comment]
Real-time updates are a nice-to-have. Whether it is with long-polling (refetching the keywords every 30s for instance), web sockets or even having a button to manually refresh the page, it would work. The core part of the challenge is not there.

The main difficulty is to split the async CSV parsing from the async scraping. Factor in that the async process requires maximum observability. Would using Node.js built-in worker threads allow us to easily know how many jobs are pending? Retry some failed background jobs? 

In most cases, it is best to go with a database-backed solution such as [Bull](https://github.com/OptimalBits/bull). It has become the more or less de factor solution, e.g., it is on the official docs for Ne**s**t.JS: https://docs.nestjs.com/techniques/queues

**Comment by nhantran3395** [comment]
I agree.

Node.js built-in worker thread will only help with offloading the tedious synchronous to another worker so that the main process can still handle incoming requests. It will not help if we want to retry failed jobs or track the pending state of each job.

In this case, using queue with [Bull](https://www.npmjs.com/package/bull) or [BullMQ](https://github.com/taskforcesh/bullmq) that uses Redis beneath would be the best solution.

**Comment by nhantran3395** [comment]
Hi @olivierobert : For this #44 issue, my plan is:

- Add a new status column in the `Keywords` table that tracks the status for each keyword.

![image](https://github.com/nhantran3395/scrapper/assets/50472067/4977a488-e139-45af-9132-266e960ecc2a)

- I will have a queue with BullMQ that each job will scrape Google, extract results, and save data back to the PostgreSQL database.
- So when a `POST uploads` comes in, its handler will parse keywords, add jobs corresponding with each keyword, and return immediately. On the UI, the user will see a new `upload` has been added to the list
- If they go to the `keywords` page either by clicking on the new `upload` from the `Uploads` list or clicking `Keywords` on the navigation bar, users will see the list of keywords along with the status. status `DONE` will have extracted result with them, `ERROR` or `PROCESSING` will show blank in those columns (total links, AdWords, total search results)
- I will trigger a new `GET /keywords` request on the keywords page every `45` seconds (will be adjusted based on the usage) to refresh the latest status of the keywords. This is done with the `useSWR` hook, which will show stale data while the request response is being waited. 

Please let me know if you have any ideas or feedback on this.

**Comment by olivierobert** [comment]
It is good to add the updated diagram along. Using an enum column is indeed the right call. 

Please clarify that part:
> add jobs corresponding with each keyword

It would be best to have more details on the queue setup job setup and how to handle failures. Most of the content is about handling the UI but the queue processing is the most complex.

**Comment by nhantran3395** [comment]
Hi @olivierobert : I would modify the `createNewUpload` handler to: 

1. Save new uploads and new keywords to DB (keyword will be saved with status `PROCESSING`)
2. Add jobs for each keyword in the CSV file to the scraper queue
3. Return with response 200

```
// createNewUploadHandler (handlers/create-upload.handler.ts)
import * as scraper from "../helpers/scraper";

async function createNewUploadHandler() {
  // ... code to validate request and read csv file
  try {
    // insert new record to "uploads" table and "keywords" table
    // the keyword records will be inserted with status PROCESSING
    await uploadRepository.createNew(user.userId, fileName, keywords);

    for (let keyword of keywords) {
      await scraper.addJob(keyword);
    }

    res.json({
      ok: true,
    });
  } catch (error) {
    console.error(error);
    next(error);
  }
}


```

The queue and worker logic will be entailed in `helpers/scraper` like the following:

```
// Scraper (helpers/scraper.ts)

import { Queue, Worker, type Job } from "bullmq";
import axios from "axios";
import UserAgent from "user-agents";

import { ProcessedKeywordResult } from "types";

import * as resultProcessor from "./result-processor";

const QUEUE_NAME = "Scraper";
const JOB_NAME = "scrape";

const scraperQueue = new Queue(`${QUEUE_NAME}`);

export async function addJob(keyword: string) {
  await scraperQueue.add(`${JOB_NAME}`, { keyword });
}

const worker = new Worker(`${QUEUE_NAME}`, async (job: Job) => {
  if (job.name === `${JOB_NAME}`) {
    const rawHtml = await scrape(job.data.keyword);
    let data = {};

    try {
      data = resultProcessor.process(rawHtml);
    } catch (error) {
      throw error; // resultProcessor will throw an error if the HTML cannot be processed (contains captcha, etc.)
    }

    return data;
  }
});

// scrape Google with the keyword and return raw HTML page
async function scrape(keyword: string) {
  const url = `https://google.com/search?hl=en&q=${encodeURIComponent(
    keyword
  )}`;

  const headers = {
    "User-Agent": new UserAgent({ deviceCategory: "desktop" }).toString(),
  };

  const option = {
    headers,
  };

  const res = await axios.get(url, option);
  return res.data;
}

worker.on("completed", (job: Job, data: ProcessedKeywordResult) => {
  // update record in "keywords" table as Done
});

worker.on("failed", (job: Job, data: ProcessedKeywordResult) => {
  // update record in "keywords" table as Errored
});

```

Each job will scrape Google and return the processed result (link count, Adwords, total count, etc.).

- On completion, I will update the already inserted keyword to `DONE` and save processed data.

- On failure, I will update the inserted keyword with `ERROR` and leave other fields `null`.

This snippet is oversimplified by not having `keywordId` in the job. The actual implementation should need to have the `keywordId` after the first insert (in `createNewUploadHandler`) and have it as data pass to the job so that upon completion or failure, that `keywordId` could be used to update the record.

For now, I picture error handling when a keyword cannot be scraped (i.e. Google blocked the request, or network error) as just updating the DB record to `ERRORED` and the user will try with that keyword in a new upload. As improvements, we could retry the job automatically, or add a retry button on the UI so users when they see an error can click on it to retry.

The process upon completion and failure to update the status of the keyword record might also failed due to network issues, and db errors. The failures should also be added to another queue to handle updating the records so that the record will not stuck in `PROCESSING` status.
